webrtc/README.md

```diff
@@ -0,0 +1,412 @@
+limits"](https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API/Using_data_channels#understanding_message_size_limits).
+Implementations MAY choose to send smaller messages, e.g. to reduce delays
+sending _flagged_ messages.
+
+### Ordering
+
+Implementations MAY expose an unordered byte stream abstraction to the user by
+overriding the default value of `ordered` `true` to `false` when creating a new
+data channel via
+[`RTCPeerConnection.createDataChannel`](https://www.w3.org/TR/webrtc/#dom-peerconnection-createdatachannel).
+
+## Connection Security
+
+Note that the below uses the message framing described in
+[multiplexing](#multiplexing).
+
+While WebRTC offers confidentiality and integrity via TLS, one still needs to
+authenticate the remote peer by its libp2p identity.
+
+After [Connection Establishment](#connection-establishment):
+
+1. _A_ opens a WebRTC datachannel.
+
+2. _A_ starts a Noise `XX` handshake using _A_'s and _B_'s libp2p identity. See
+   [noise-libp2p](https://github.com/libp2p/specs/tree/master/noise).
+
+   Instead of exchanging the TLS certificate fingerprints on the established
+   Noise channel once the Noise handshake succeeded, _A_ and _B_ use the [Noise
+   Prologue](https://noiseprotocol.org/noise.html#prologue) mechanism, thus
+   saving one round trip.
+
+   More specifically _A_ and _B_ set the Noise _Prologue_ to
+   `libp2p-webrtc-noise:<FINGERPRINTS>` before starting the actual Noise
+   handshake. `<FINGERPRINTS>` is the concatenation of the two TLS
+   fingerprints of _A_ and _B_ in their multihash byte representation, sorted in
+   ascending order.
+
+3. On success of the authentication handshake, the used datachannel is
+   closed and the plain WebRTC connection is used with its multiplexing
+   capabilities via datachannels. See [Multiplexing](#multiplexing).
+
+Note: WebRTC supports different hash functions to hash the TLS certificate (see
+https://datatracker.ietf.org/doc/html/rfc8122#section-5). The hash function used
+in WebRTC and the hash function used in the multiaddr `/certhash` component MUST
+be the same. On mismatch the final Noise handshake MUST fail.
+
+### Open Questions
+
+- Can a _Browser_ access the fingerprint of its TLS certificate?
+
+  Chrome allows you to access the fingerprint of any locally-created certificate
+  directly via `RTCCertificate#getFingerprints`. Firefox does not allow you to
+  do so. Browser compatibility can be found
+  [here](https://developer.mozilla.org/en-US/docs/Web/API/RTCCertificate). In
+  practice, this is not an issue since the fingerprint is embedded in the local
+  SDP string.
+
+- Is the above proposed additional handshake secure? See also alternative
+  proposed Handshake for
+  [WebTransport](https://github.com/libp2p/specs/pull/404).
+
+- Would it be more efficient for _B_ to initiate the Noise handshake? In other
+  words, who is able to write on an established WebRTC connection first? _A_ or
+  _B_?
```

Comment:
If I am not missing something WebRTC's connection establishment finishes with the DTLS (modification of TLS 1.2) handshake. As far as I know, in a TLS 1.2 handshake the server side can send application data first right after having send the TLS `finished` message.

With the above in mind, we could safe 0.5 RTT by having the server initiate the Noise handshake. The nice side-effect would be that the connection initiator (here `A`, aka the browser) would be able to send application data first, as it sends the last Noise handshake message.

What do folks think? @ckousik in case I am not missing something, could you validate the above using the js-libp2p and go-libp2p implementation?

Reply:
I don't think this will work since the server does not know the initiator's peerId until completion of the Noise handshake.

Reply:
@ckousik is right. The initiator of the Noise handshake is the one who finishes the handshake first, and can send application data first.  This is in contrast with the TLS 1.3 handshake, where the server can send 0.5-RTT data.

Reply:
> The server will not be able to initiate the handshake without knowing the remote's peerId.

Why does the Noise initiator (here the server) need to know the peer ID of the Noise responder (here the client or browser)? If I am not mistaken this is not required by the Noise XX handshake.

Reply:
> The initiator of the Noise handshake is the one who finishes the handshake first

Ah, mistake on my end. Correct.

The above suggestion would still be an optimization, i.e. it would allow the server (webrtc connection responder, here Noise handshake initiator) to send application data 0.5 RTT earlier. Things would stay as is for the client (webrtc connection initiator, here Noise handshake responder).

Reply:
@ckousik @marten-seemann how much work do you expect this would be on the go-libp2p side?

@John-LittleBearLabs @ckousik @achingbrain  how much work do you expect this would be on the js-libp2p side?

I expect this to be a small change only on the rust-libp2p side.

Reply:
> The above suggestion would still be an optimization, i.e. it would allow the server (webrtc connection responder, here Noise handshake initiator) to send application data 0.5 RTT earlier. Things would stay as is for the client (webrtc connection initiator, here Noise handshake responder).

what's the application data here? is it some Noise term? afaik the client drives communication (i.e. after noise is finished, it opens new substreams according to the inner protocol), so I'm not sure why server should be the first sending app data.

Reply:
The server could e.g. exchange identify information or open a gossipsub stream.

@melekes if I am not mistaken, this should be a simple change on the Rust side, i.e. just a change of roles. Am I missing something?

Reply:
Ah, okay. Yeah, it should be very easy to change in Rust impl.

Reply:
On the go side, we need to change the roles and modify this https://github.com/libp2p/go-libp2p/blob/164ccd79587daf8adb0cbb6fa49b433193d46521/p2p/security/noise/handshake.go#L275-L281

Reply:
Documented in b5466fd. I am resolving here. Thanks for the input everyone.